File: src-cpp/leetcode/two-sum.cpp
#include <bits/stdc++.h>
using namespace std;

// 1 pass
vector<int> twoSum(vector<int> &nums, int target)
{
    unordered_map<int, int> m; // num->idx

    for (int i = 0; i < nums.size(); i++)
    {
        if (m.find(target - nums[i]) != m.end())
            return {i, m[target - nums[i]]};

        m[nums[i]] = i;
    }

    return {};
}

// 2 pass
vector<int> twoSum(vector<int> &nums, int target)
{
    unordered_map<int, int> m; // num->idx

    for (int i = 0; i < nums.size(); i++)
        m[nums[i]] = i;

    for (int i = 0; i < nums.size(); i++)
        if (m.find(target - nums[i]) != m.end() && m[target - nums[i]] != i)
            return {i, m[target - nums[i]]};

    return {};
}

File: src-ts/leetcode/two-sum.ts
// 1 pass
function twoSum(nums: number[], target: number): number[] {
  const map: { [key: number]: number } = {}; // idx->num

  for (const [index, num] of nums.entries()) {
    if (map[target - num] !== undefined) {
      return [map[target - num], index];
    }

    map[num] = index;
  }

  return [];
}
